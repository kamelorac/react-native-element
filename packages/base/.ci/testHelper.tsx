import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
