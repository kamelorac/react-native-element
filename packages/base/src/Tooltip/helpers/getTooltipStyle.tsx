import { I18nManager, StyleSheet, ViewStyle } from 'react-native';
import { ScreenWidth, ScreenHeight } from '../../helpers';
import getTooltipCoordinate from './getTooltipCoordinate';

export const getTooltipStyle = ({
  yOffset,
  xOffset,
  elementHeight,
  elementWidth,
  width,
  height,
  withPointer,
  backgroundColor,
  containerStyle,
}): ViewStyle => {
  const { x, y } = getTooltipCoordinate(
    xOffset,
    yOffset,
    elementWidth,
    elementHeight,
    ScreenWidth,
    ScreenHeight,
    width,
    height,
    withPointer
  );
  return StyleSheet.flatten([
    {
      position: 'absolute',
      [I18nManager.isRTL ? 'right' : 'left']: x,
      top: y,
      width,
      height,
      backgroundColor,
      // default styles
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      flex: 1,
      borderRadius: 10,
      padding: 10,
    },
    containerStyle,
  ]);
};
