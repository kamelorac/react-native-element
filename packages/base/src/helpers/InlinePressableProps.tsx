import { PressableProps, GestureResponderEvent } from 'react-native';

export interface InlinePressableProps {
  /**
   * Called when a single tap gesture is detected.
   * @type GestureResponderEventHandler
   */
  onPress?: (event: GestureResponderEvent) => void;

  /**
   * Called when a touch is engaged before `onPress`.
   * @type GestureResponderEventHandler
   */
  onPressIn?: (event: GestureResponderEvent) => void;

  /**
   * Called when a touch is released before `onPress`.
   * @type GestureResponderEventHandler
   */
  onPressOut?: (event: GestureResponderEvent) => void;

  /**
   * Called when a long-tap gesture is detected.
   * @type GestureResponderEventHandler
   */
  onLongPress?: (event: GestureResponderEvent) => void;

  /**
   * @default None
   * @type PressableProps except click handlers
   */
  pressableProps?: Omit<
    PressableProps,
    'onPress' | 'onLongPress' | 'onPressIn' | 'onPressOut'
  >;
}
