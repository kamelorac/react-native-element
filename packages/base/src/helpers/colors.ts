import { StyleSheet } from 'react-native';

interface PlatformColors {
  primary: string;
  secondary: string;
  grey: string;
  searchBg: string;
  success: string;
  error: string;
  warning: string;
}

export interface Colors {
  readonly primary: string;
  readonly secondary: string;
  readonly background: string;
  readonly white: string;
  readonly black: string;
  readonly grey0: string;
  readonly grey1: string;
  readonly grey2: string;
  readonly grey3: string;
  readonly grey4: string;
  readonly grey5: string;
  readonly greyOutline: string;
  readonly searchBg: string;
  readonly success: string;
  readonly warning: string;
  readonly error: string;
  readonly disabled: string;
  readonly divider: string;
  readonly platform: {
    ios: PlatformColors;
    android: PlatformColors;
    web: PlatformColors;
    default: PlatformColors;
  };
}

export const lightColors: Colors = {
  primary: '#2089dc',
  secondary: '#ad1457',
  background: '#ffffff',
  white: '#ffffff',
  black: '#242424',
  grey0: '#393e42',
  grey1: '#43484d',
  grey2: '#5e6977',
  grey3: '#86939e',
  grey4: '#bdc6cf',
  grey5: '#e1e8ee',
  greyOutline: '#bbb',
  searchBg: '#303337',
  success: '#52c41a',
  error: '#ff190c',
  warning: '#faad14',
  disabled: 'hsl(208, 8%, 90%)',
  // Darker color if hairlineWidth is not thin enough
  divider: StyleSheet.hairlineWidth < 1 ? '#bcbbc1' : 'rgba(0, 0, 0, 0.12)',
  platform: {
    ios: {
      primary: '#007aff',
      secondary: '#5856d6',
      grey: '#7d7d7d',
      searchBg: '#dcdce1',
      success: '#4cd964',
      error: '#ff3b30',
      warning: '#ffcc00',
    },
    android: {
      primary: '#2196f3',
      secondary: '#9C27B0',
      grey: 'rgba(0, 0, 0, 0.54)',
      searchBg: '#dcdce1',
      success: '#4caf50',
      error: '#f44336',
      warning: '#ffeb3b',
    },
    web: {
      primary: '#2089dc',
      secondary: '#ca71eb',
      grey: '#393e42',
      searchBg: '#303337',
      success: '#52c41a',
      error: '#ff190c',
      warning: '#faad14',
    },
    default: {
      primary: '#007aff',
      secondary: '#5856d6',
      grey: '#7d7d7d',
      searchBg: '#dcdce1',
      success: '#4cd964',
      error: '#ff3b30',
      warning: '#ffcc00',
    },
  },
};

export const darkColors: Colors = {
  primary: '#439ce0',
  secondary: '#aa49eb',
  background: '#080808',
  white: '#080808',
  black: '#f2f2f2',
  grey5: '#393e42',
  grey4: '#43484d',
  grey3: '#5e6977',
  grey2: '#86939e',
  grey1: '#bdc6cf',
  grey0: '#e1e8ee',
  greyOutline: '#bbb',
  searchBg: '#303337',
  success: '#439946',
  error: '#bf2c24',
  warning: '#cfbe27',
  disabled: 'hsl(208, 8%, 90%)',
  // Darker color if hairlineWidth is not thin enough
  divider: StyleSheet.hairlineWidth < 1 ? '#84838a' : 'rgba(0, 0, 0, 0.12)',
  platform: {
    ios: {
      primary: '#1b262c',
      secondary: '#2089dc',
      grey: '#ffffff',
      searchBg: '#393e42',
      success: '#439946',
      error: '#bf2c24',
      warning: '#cfbe27',
    },
    android: {
      primary: '#1b262c',
      secondary: '#2089dc',
      grey: '#393e42',
      searchBg: '#393e42',
      success: '#439946',
      error: '#bf2c24',
      warning: '#cfbe27',
    },
    web: {
      primary: '#1b262c',
      secondary: '#2089dc',
      grey: '#ffffff',
      searchBg: '#393e42',
      success: '#439946',
      error: '#bf2c24',
      warning: '#cfbe27',
    },
    default: {
      primary: '#1b262c',
      secondary: '#2089dc',
      grey: '#ffffff',
      searchBg: '#393e42',
      success: '#439946',
      error: '#bf2c24',
      warning: '#cfbe27',
    },
  },
};
