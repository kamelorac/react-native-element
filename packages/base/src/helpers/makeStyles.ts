import { useMemo } from 'react';
import { StyleSheet } from 'react-native';

export const makeStyles =
  <P, T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>>(
    styles: T | ((props: P) => T)
  ) =>
  (props?: P): T => {
    return useMemo(() => {
      return StyleSheet.create(
        typeof styles === 'function' ? styles(props) : styles
      );
    }, [props]);
  };
