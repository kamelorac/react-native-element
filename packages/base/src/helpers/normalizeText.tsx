import { moderateScale } from 'react-native-size-matters';

function normalize(number: number, factor = 0.25) {
  return moderateScale(number, factor);
}

export default normalize;
