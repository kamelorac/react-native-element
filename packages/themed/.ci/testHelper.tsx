import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
