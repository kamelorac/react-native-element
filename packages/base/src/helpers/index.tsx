import { Platform, Dimensions, PressableProps, ColorValue } from 'react-native';
import color from 'color';
import renderNode from './renderNode';
import getIconType, { registerCustomIconType } from './getIconType';
import normalizeText from './normalizeText';
import { Colors, lightColors, darkColors } from './colors';
import { InlinePressableProps } from './InlinePressableProps';
import React from 'react';

const Screen = Dimensions.get('window');
const ScreenWidth = Screen.width;
const ScreenHeight = Screen.height;
const isIOS = Platform.OS === 'ios';
const getBehaviorType = Platform.OS === 'ios' ? 'padding' : 'height';

export type StringOmit<K extends string> = K | Omit<string, K>;

export type RneFunctionComponent<T> = React.FunctionComponent<
  T & {
    theme?: Theme;
    children?: React.ReactNode | undefined;
  }
>;

export interface ThemeSpacing {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  xl: number;
}

export const defaultSpacing = { xs: 2, sm: 4, md: 8, lg: 12, xl: 24 };

export const defaultTheme: Theme = {
  colors: lightColors,
  spacing: defaultSpacing,
};

export type Theme = {
  colors: Colors;
  spacing: ThemeSpacing;
};

export const androidRipple = (
  rippleColor?: string | ColorValue
): PressableProps['android_ripple'] => {
  return {
    borderless: false,
    color: rippleColor,
    radius: -5,
  };
};

export const patchWebProps = <T extends Record<any, any>>({
  updateTheme,
  replaceTheme,
  onClear,
  ...rest
}: T) => {
  return rest;
};

export type { Colors, InlinePressableProps };

export {
  renderNode,
  getIconType,
  normalizeText,
  ScreenWidth,
  ScreenHeight,
  isIOS,
  lightColors,
  darkColors,
  color,
  getBehaviorType,
  registerCustomIconType,
};

export { default as BackgroundImage } from './BackgroundImage';
export { default as fonts } from './fonts';

export { makeStyles } from './makeStyles';
