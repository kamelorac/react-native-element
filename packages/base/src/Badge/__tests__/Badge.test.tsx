import React from 'react';
import { Badge } from '..';
import { Text, TouchableWithoutFeedback, Pressable, View } from 'react-native';
import { renderWithWrapper } from '../../../.ci/testHelper';
import { fireEvent } from '@testing-library/react-native';
import { lightColors } from '../../helpers';

describe('Badge Component', () => {
  it('should match snapshot', () => {
    const component = renderWithWrapper(<Badge value={10} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should work with value prop as ReactElement', () => {
    const component = renderWithWrapper(<Badge value={<Text>foo</Text>} />);
    expect(component.queryByText('foo')).toBeTruthy();
  });

  it('should work with value prop as string', () => {
    const component = renderWithWrapper(<Badge value="foo" />);
    expect(component.queryByText('foo')).toBeTruthy();
  });

  it('should apply container style in the badge', () => {
    const { wrapper } = renderWithWrapper(
      <Badge value={10} containerStyle={{ backgroundColor: 'orange' }} />
    );
    expect(wrapper.findByType(View).props.style.backgroundColor).toBe('orange');
  });

  it('should allow badge style', () => {
    const { wrapper } = renderWithWrapper(
      <Badge value={10} badgeStyle={{ backgroundColor: 'pink' }} />,
      'RNE__Badge'
    );
    expect(wrapper.props.style.backgroundColor).toBe('pink');
  });

  it('should allow custom component', () => {
    const { wrapper } = renderWithWrapper(
      <Badge value={10} Component={TouchableWithoutFeedback} />
    );
    expect(wrapper.findAllByType(TouchableWithoutFeedback).length).toBe(1);
  });

  it('should have a touchable when onPress is passed in', () => {
    const handler = jest.fn();
    const { wrapper } = renderWithWrapper(
      <Badge value={10} onPress={handler} />
    );
    const component = wrapper.findByType(Pressable);
    fireEvent.press(component);
    expect(handler).toBeCalledTimes(1);
  });

  it('should use text props', () => {
    const { getAllByTestId } = renderWithWrapper(
      <Badge textProps={{ testID: 'text' }} value={10} />
    );
    const elements = getAllByTestId('text');
    expect(elements).toHaveLength(1);
  });

  describe('Mini badge', () => {
    it.each`
      status
      ${'primary'}
      ${'success'}
      ${'warning'}
      ${'error'}
    `('accepts $status', ({ status }) => {
      const { wrapper } = renderWithWrapper(
        <Badge status={status} />,
        'RNE__Badge'
      );
      expect(wrapper.props.style.backgroundColor).toBe(lightColors[status]);
    });
  });
});
