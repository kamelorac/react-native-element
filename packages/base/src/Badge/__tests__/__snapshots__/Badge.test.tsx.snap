// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    style={{}}
    testID="RNE__Badge__Container"
  >
    <View
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#2089dc",
          "borderColor": "#fff",
          "borderRadius": 9,
          "borderWidth": 0.5,
          "height": 18,
          "justifyContent": "center",
          "minWidth": 18,
        }
      }
      testID="RNE__Badge"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 12,
            "paddingHorizontal": 4,
          }
        }
      >
        10
      </Text>
    </View>
  </View>
</View>
`;
